nx/source/services/bsd.c

```diff
@@ -92,30 +121,54 @@ static Result _bsdStartMonitor(Service* srv, u64 pid) {
     return rc;
 }
 
-Result bsdInitialize(TransferMemory* tmem) {
+const BsdConfig *bsdGetDefaultConfig(void) {
+    return &g_defaultBsdConfig;
+}
+
+Result bsdInitialize(const BsdConfig *config) {
     const char* bsd_srv = "bsd:s";
+
+    if(serviceIsActive(&g_bsdSrv) || serviceIsActive(&g_bsdMonitor))
+        return MAKERESULT(Module_Libnx, LibnxError_AlreadyInitialized);
+
     Result rc = smGetService(&g_bsdSrv, bsd_srv);
 
     if (R_FAILED(rc)) {
         bsd_srv = "bsd:u";
         rc = smGetService(&g_bsdSrv, bsd_srv);
     }
+    if(R_FAILED(rc)) goto error;
 
-    if (R_SUCCEEDED(rc)) {
-        rc = smGetService(&g_bsdMonitor, bsd_srv);
+    rc = smGetService(&g_bsdMonitor, bsd_srv);
+    if(R_FAILED(rc)) goto error;
 
-        if (R_SUCCEEDED(rc)) {
-            rc = _bsdRegisterClient(&g_bsdSrv, tmem, &g_bsdClientPid);
+    rc = tmemCreate(&g_bsdTmem, _bsdGetTransferMemSizeForConfig(config), 0);
+    if(R_FAILED(rc)) goto error;
 
-            if (R_SUCCEEDED(rc)) {
-                rc = _bsdStartMonitor(&g_bsdMonitor, g_bsdClientPid);
-            }
-        }
-    }
+    rc = _bsdRegisterClient(&g_bsdSrv, &g_bsdTmem, config, &g_bsdClientPid);
+    if(R_FAILED(rc)) goto error;
 
+    rc = _bsdStartMonitor(&g_bsdMonitor, g_bsdClientPid);
+    if(R_FAILED(rc)) goto error;
+
+    return rc;
+
+error:
+    if(serviceIsActive(&g_bsdMonitor))
```

Comment:
serviceClose() was designed so that you don't need to check if it's active

Reply:
Agree